src/main/scala/uk/gov/nationalarchives/consignmentexport/S3Files.scala

```diff
@@ -5,32 +5,47 @@ import java.util.UUID
 
 import cats.effect.IO
 import cats.implicits._
+import com.typesafe.config.ConfigFactory
 import org.typelevel.log4cats.SelfAwareStructuredLogger
+import software.amazon.awssdk.services.s3.model.GetObjectResponse
 import uk.gov.nationalarchives.aws.utils.S3Utils
+import uk.gov.nationalarchives.consignmentexport.Config.Configuration
 import uk.gov.nationalarchives.consignmentexport.Main.directoryType
 import uk.gov.nationalarchives.consignmentexport.Utils._
 import uk.gov.nationalarchives.consignmentexport.Validator.ValidatedFileMetadata
-import scala.concurrent.duration._
 
+import scala.concurrent.duration._
 import scala.language.postfixOps
 
-class S3Files(s3Utils: S3Utils)(implicit val logger: SelfAwareStructuredLogger[IO]) {
+class S3Files(s3Utils: S3Utils, config: Configuration)(implicit val logger: SelfAwareStructuredLogger[IO]) {
+  private val downloadBatchSize = config.s3.downloadFilesBatchSize
+  private val downloadDelay = config.s3.downloadBatchDelayMs
 
-  def createDownloadDirectories(files: List[ValidatedFileMetadata], consignmentReference: String, rootLocation: String) = {
+  def createDownloadDirectories(files: List[ValidatedFileMetadata], consignmentReference: String, rootLocation: String): IO[List[Boolean]] = {
     IO {
       new File(s"$rootLocation/$consignmentReference").mkdirs()
       files.filter(_.fileType == "Folder").map(f => new File(s"$rootLocation/$consignmentReference/${f.clientSideOriginalFilePath}").mkdirs())
     }
   }
 
-  def downloadFiles(files: List[ValidatedFileMetadata], bucket: String, consignmentId: UUID, consignmentReference: String, rootLocation: String): IO[Unit] = for {
-    _ <- createDownloadDirectories(files, consignmentReference, rootLocation)
-    _ <- files.filter(_.fileType != directoryType).map(file => {
-      IO.sleep(100 milliseconds).flatMap(_ =>
-        s3Utils.downloadFiles(bucket, s"$consignmentId/${file.fileId}", s"$rootLocation/$consignmentReference/${file.clientSideOriginalFilePath}".toPath.some))
-    }).sequence
-    _ <- logger.info(s"Files downloaded from S3 for consignment $consignmentId")
-  } yield ()
+  def batchDownloadingFiles(batch: List[ValidatedFileMetadata], bucket: String, consignmentId: UUID, consignmentReference: String, rootLocation: String): IO[List[GetObjectResponse]] = {
+    IO.sleep(downloadDelay milliseconds).flatMap(_ => {
```

Comment:
I'm getting a bit lost with these sleeps but do you need to put `Thread.sleep` in here rather than `IO.sleep` to stop the weird thread errors?.

Reply:
Probably. Will see if the `Thread.sleep` fixes that thread error first.

Reply:
Switching to Thread.sleep hasn't helped with the thread shut down issue.